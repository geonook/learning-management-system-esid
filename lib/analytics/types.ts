/**
 * Analytics Type Definitions for Primary School LMS
 * Comprehensive types for learning analytics and reporting system
 */

// Base analytics interfaces
export interface AnalyticsTimeRange {
  startDate: string // ISO date string
  endDate: string   // ISO date string
  period: 'day' | 'week' | 'month' | 'semester' | 'year' | 'custom'
}

export interface AnalyticsFilters {
  grades?: number[]
  tracks?: ('local' | 'international')[]
  classes?: string[]
  teachers?: string[]
  courseTypes?: ('LT' | 'IT' | 'KCFS')[]
  students?: string[]
  timeRange: AnalyticsTimeRange
}

// Student Analytics Types
export interface StudentLearningMetrics {
  studentId: string
  studentName: string
  grade: number
  track: 'local' | 'international'
  classId: string
  className: string
  
  // Academic Performance
  overallAverage: number | null
  formativeAverage: number | null
  summativeAverage: number | null
  semesterGrade: number | null
  
  // Progress Metrics
  improvementRate: number // Percentage improvement over time
  consistency: number     // Score variance metric (lower = more consistent)
  engagement: number      // Participation/submission rate
  
  // Risk Indicators
  atRisk: boolean
  riskFactors: string[]
  interventionNeeded: boolean
  
  // Trends
  performanceTrend: 'improving' | 'declining' | 'stable'
  recentScores: number[]
  timeStamps: string[]
}

export interface StudentProgressTimeline {
  studentId: string
  assessments: Array<{
    examId: string
    examName: string
    examDate: string
    score: number
    assessmentCode: string
    courseType: 'LT' | 'IT' | 'KCFS'
    percentile: number // Student's rank within class
  }>
  milestones: Array<{
    date: string
    event: 'grade_improvement' | 'consistent_performance' | 'achievement' | 'concern'
    description: string
    impact: 'positive' | 'negative' | 'neutral'
  }>
}

// Teacher Analytics Types
export interface TeacherPerformanceMetrics {
  teacherId: string
  teacherName: string
  teacherType: 'LT' | 'IT' | 'KCFS'
  
  // Class Performance
  classesCount: number
  studentsCount: number
  averageClassPerformance: number
  passRate: number // Percentage of students passing (>= 70%)
  
  // Teaching Effectiveness
  studentImprovementRate: number // Average improvement across all students
  consistencyScore: number       // How consistent are student outcomes
  engagementRate: number          // Assignment submission rate
  
  // Workload Metrics
  examsCreated: number
  scoresEntered: number
  averageGradingTime: number // Days between exam and score entry
  
  // Comparative Metrics
  gradeRankPercentile: number    // Performance vs other teachers in same grade
  subjectRankPercentile: number  // Performance vs other teachers of same type
  
  // Trend Data
  monthlyPerformance: Array<{
    month: string
    averageScore: number
    passRate: number
    studentsCount: number
  }>
}

export interface ClassComparisonMetrics {
  classId: string
  className: string
  grade: number
  track: 'local' | 'international'
  studentsCount: number
  
  // Performance Metrics
  averageScore: number
  medianScore: number
  standardDeviation: number
  passRate: number
  
  // Course-specific Performance
  ltPerformance: number | null
  itPerformance: number | null
  kcfsPerformance: number | null
  
  // Ranking
  gradeRank: number    // Rank within same grade
  schoolRank: number   // Rank within entire school
  trackRank: number    // Rank within same track
  
  // Distribution
  scoreDistribution: Array<{
    range: string // e.g., "90-100", "80-89"
    count: number
    percentage: number
  }>
}

// Administrative Analytics Types
export interface SchoolOverviewMetrics {
  totalStudents: number
  totalTeachers: number
  totalClasses: number
  totalExams: number
  
  // Overall Performance
  schoolAverageScore: number
  overallPassRate: number
  improvementRate: number
  
  // Grade Performance
  gradePerformance: Array<{
    grade: number
    studentsCount: number
    averageScore: number
    passRate: number
    trend: 'improving' | 'declining' | 'stable'
  }>
  
  // Track Performance
  trackPerformance: Array<{
    track: 'local' | 'international'
    studentsCount: number
    averageScore: number
    passRate: number
    classesCount: number
  }>
  
  // Teacher Performance Summary
  teacherSummary: Array<{
    teacherType: 'LT' | 'IT' | 'KCFS'
    teachersCount: number
    averageClassPerformance: number
    totalStudents: number
  }>
  
  // Recent Trends
  monthlyTrends: Array<{
    month: string
    averageScore: number
    passRate: number
    examCount: number
    alertCount: number
  }>
}

// Predictive Analytics Types
export interface LearningPrediction {
  studentId: string
  predictionType: 'semester_grade' | 'pass_likelihood' | 'improvement_potential'
  confidence: number // 0-1, how confident the model is
  prediction: number
  factors: Array<{
    factor: string
    impact: number    // -1 to 1, negative means hurting performance
    importance: number // 0-1, how important this factor is
  }>
  recommendations: string[]
  dataPoints: number // How many scores/assessments were used
}

export interface RiskAssessment {
  studentId: string
  riskLevel: 'low' | 'medium' | 'high' | 'critical'
  riskScore: number // 0-100
  primaryConcerns: string[]
  interventions: Array<{
    type: 'academic_support' | 'attendance_monitoring' | 'parent_conference' | 'peer_tutoring'
    priority: 'high' | 'medium' | 'low'
    description: string
    estimatedImpact: number // Expected improvement percentage
  }>
  timeline: {
    identifiedDate: string
    lastAssessment: string
    nextReview: string
  }
}

// Visualization Data Types
export interface ChartDataPoint {
  x: number | string
  y: number
  label?: string
  color?: string
  metadata?: Record<string, any>
}

export interface TrendAnalysis {
  period: string
  data: ChartDataPoint[]
  trend: 'up' | 'down' | 'stable'
  changePercentage: number
  significance: 'high' | 'medium' | 'low' // Statistical significance
}

export interface PerformanceDistribution {
  ranges: Array<{
    min: number
    max: number
    range: string        // Format: "90-100", "80-89", etc.
    count: number
    percentage: number
    label: string
  }>
  mean: number
  median: number
  standardDeviation: number
  skewness: number // -1 to 1, indicates distribution shape
}

// Export all analytics types
export interface AnalyticsReport {
  id: string
  title: string
  type: 'student' | 'teacher' | 'class' | 'school' | 'predictive'
  generatedAt: string
  generatedBy: string
  filters: AnalyticsFilters
  data: StudentLearningMetrics[] | TeacherPerformanceMetrics[] | ClassComparisonMetrics[] | SchoolOverviewMetrics
  summary: {
    keyInsights: string[]
    recommendations: string[]
    dataQuality: 'excellent' | 'good' | 'fair' | 'poor'
    sampleSize: number
  }
}

// Query Builder Types
export interface AnalyticsQuery {
  queryId: string
  name: string
  description: string
  type: 'aggregation' | 'comparison' | 'trend' | 'prediction'
  sql?: string // Raw SQL for complex queries
  parameters: Record<string, any>
  cacheDuration: number // Minutes to cache results
  permissions: ('admin' | 'head' | 'teacher')[]
}

export interface AnalyticsCache {
  queryId: string
  result: any
  generatedAt: string
  expiresAt: string
  filters: AnalyticsFilters
  hitCount: number
}