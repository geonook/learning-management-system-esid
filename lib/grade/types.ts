import { z } from 'zod'

// Assessment codes - these are the canonical identifiers for calculations
export const ASSESSMENT_CODES = [
  'FA1', 'FA2', 'FA3', 'FA4', 'FA5', 'FA6', 'FA7', 'FA8',
  'SA1', 'SA2', 'SA3', 'SA4',
  'FINAL'
] as const

export type AssessmentCode = typeof ASSESSMENT_CODES[number]

export const FORMATIVE_CODES: AssessmentCode[] = ['FA1', 'FA2', 'FA3', 'FA4', 'FA5', 'FA6', 'FA7', 'FA8']
export const SUMMATIVE_CODES: AssessmentCode[] = ['SA1', 'SA2', 'SA3', 'SA4']
export const FINAL_CODES: AssessmentCode[] = ['FINAL']

// Zod schemas for validation
export const AssessmentCodeSchema = z.enum(ASSESSMENT_CODES)

export const ScoreSchema = z.number().min(0).max(100).nullable()

export const ScoresMapSchema = z.record(
  AssessmentCodeSchema,
  ScoreSchema
)

export const GradeCalculationInputSchema = z.object({
  scores: ScoresMapSchema,
  studentId: z.string(),
  classId: z.string(),
  examId: z.string().optional()
})

// TypeScript types
export type ScoreValue = number | null
export type ScoresMap = Record<AssessmentCode, ScoreValue>

export interface GradeCalculationInput {
  scores: ScoresMap
  studentId: string
  classId: string
  examId?: string
}

export interface GradeCalculationResult {
  formativeAvg: number | null
  summativeAvg: number | null
  semesterGrade: number | null
  totalScoresUsed: number
  formativeScoresUsed: number
  summativeScoresUsed: number
  finalScoreUsed: boolean
}

// Assessment display name override types
export interface AssessmentDisplayOverride {
  code: AssessmentCode
  displayName: string
  context: 'class' | 'grade_track' | 'default'
  classId?: string
  grade?: number
  track?: string
}