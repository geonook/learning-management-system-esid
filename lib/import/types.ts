/**
 * Data Import Types and Schemas for LMS-ESID
 * Defines the structure and validation for CSV data import
 */

import { z } from 'zod'

// Import data types matching database schema
export const UserImportSchema = z.object({
  email: z.string().email('Invalid email format'),
  full_name: z.string().min(1, 'Full name is required'),
  role: z.enum(['admin', 'head', 'teacher'], {
    errorMap: () => ({ message: 'Role must be admin, head, or teacher' })
  }),
  teacher_type: z.enum(['LT', 'IT', 'KCFS']).optional().nullable(),
  grade: z.number().int().min(1).max(6).optional().nullable(),
  track: z.enum(['local', 'international']).optional().nullable(),
  is_active: z.boolean().default(true)
})

export const ClassImportSchema = z.object({
  name: z.string().min(1, 'Class name is required').regex(
    /^G[1-6] (Trailblazers|Discoverers|Adventurers|Innovators|Explorers|Navigators|Inventors|Voyagers|Pioneers|Guardians|Pathfinders|Seekers|Visionaries|Achievers)$/,
    'Class name must follow format: G[1-6] [StandardClassName]'
  ),
  grade: z.number().int().min(1).max(6, 'Grade must be between 1 and 6'),
  level: z.enum(['E1', 'E2', 'E3'], {
    errorMap: () => ({ message: 'Level must be E1, E2, or E3' })
  }),
  track: z.enum(['local', 'international'], {
    errorMap: () => ({ message: 'Track must be local or international' })
  }),
  teacher_email: z.string().email('Invalid teacher email').optional().nullable(),
  academic_year: z.string().regex(/^\d{4}$/, 'Academic year must be YYYY format').default('2024'),
  is_active: z.boolean().default(true)
})

export const StudentImportSchema = z.object({
  student_id: z.string().min(1, 'Student ID is required'),
  full_name: z.string().min(1, 'Full name is required'),
  grade: z.number().int().min(1).max(6, 'Grade must be between 1 and 6'),
  level: z.enum(['E1', 'E2', 'E3'], {
    errorMap: () => ({ message: 'Level must be E1, E2, or E3' })
  }).optional().nullable(), // Inherited from class, but can be specified
  track: z.enum(['local', 'international'], {
    errorMap: () => ({ message: 'Track must be local or international' })
  }),
  class_name: z.string().optional().nullable(), // Will be mapped to class_id
  is_active: z.boolean().default(true)
})

export const ScoreImportSchema = z.object({
  student_id: z.string().min(1, 'Student ID is required'), // External student ID
  exam_name: z.string().min(1, 'Exam name is required'),
  assessment_code: z.enum([
    'FA1', 'FA2', 'FA3', 'FA4', 'FA5', 'FA6', 'FA7', 'FA8',
    'SA1', 'SA2', 'SA3', 'SA4', 'FINAL'
  ], {
    errorMap: () => ({ message: 'Invalid assessment code' })
  }),
  score: z.number().min(0).max(100, 'Score must be between 0 and 100'),
  entered_by_email: z.string().email('Invalid teacher email') // Will be mapped to entered_by UUID
})

// Inferred types
export type UserImport = z.infer<typeof UserImportSchema>
export type ClassImport = z.infer<typeof ClassImportSchema>
export type StudentImport = z.infer<typeof StudentImportSchema>
export type ScoreImport = z.infer<typeof ScoreImportSchema>

// Import validation result types
export interface ImportValidationResult<T> {
  valid: T[]
  invalid: ImportValidationError[]
  summary: {
    total: number
    valid: number
    invalid: number
    validPercent: number
  }
}

export interface ImportValidationError {
  row: number
  data: Record<string, any>
  errors: string[]
}

// Import execution result types
export interface ImportExecutionResult {
  success: boolean
  summary: {
    users: { created: number; updated: number; errors: number }
    classes: { created: number; updated: number; errors: number }
    students: { created: number; updated: number; errors: number }
    scores: { created: number; updated: number; errors: number }
  }
  errors: ImportExecutionError[]
  warnings: ImportExecutionWarning[]
}

export interface ImportExecutionError {
  stage: 'users' | 'classes' | 'students' | 'scores'
  operation: 'create' | 'update'
  data: Record<string, any>
  error: string
}

export interface ImportExecutionWarning {
  stage: 'users' | 'classes' | 'students' | 'scores'
  message: string
  data?: Record<string, any>
}

// CSV parsing options
export interface CSVParseOptions {
  delimiter?: string
  skipEmptyLines?: boolean
  skipLinesWithError?: boolean
  encoding?: 'utf8' | 'utf16le' | 'latin1'
  maxRows?: number
}

// Import stage configuration
export interface ImportStageConfig {
  stage: 'users' | 'classes' | 'students' | 'scores'
  schema: z.ZodSchema<any>
  requiredColumns: string[]
  optionalColumns: string[]
  dependencies: string[] // Which stages must complete first
}

// File upload types
export interface ImportFileInfo {
  name: string
  size: number
  type: string
  stage: 'users' | 'classes' | 'students' | 'scores'
  content: string
}

// Import session tracking
export interface ImportSession {
  id: string
  created_at: string
  created_by: string
  status: 'pending' | 'validating' | 'importing' | 'completed' | 'failed'
  files: ImportFileInfo[]
  validation_results: Record<string, ImportValidationResult<any>>
  execution_result?: ImportExecutionResult
  progress: {
    current_stage: string
    completed_stages: string[]
    total_progress: number // 0-100
  }
}

// Common CSV column mappings
export const CSV_COLUMN_MAPPINGS = {
  users: {
    email: ['email', 'Email', 'EMAIL', 'teacher_email', 'Teacher Email'],
    full_name: ['full_name', 'Full Name', 'name', 'Name', 'teacher_name'],
    role: ['role', 'Role', 'ROLE', 'position'],
    teacher_type: ['teacher_type', 'Teacher Type', 'type', 'Type'],
    grade: ['grade', 'Grade', 'GRADE'],
    track: ['track', 'Track', 'TRACK', 'stream']
  },
  classes: {
    name: ['name', 'Name', 'class_name', 'Class Name', 'class'],
    grade: ['grade', 'Grade', 'GRADE'],
    level: ['level', 'Level', 'LEVEL', 'class_level', 'performance_level'],
    track: ['track', 'Track', 'TRACK', 'stream'],
    teacher_email: ['teacher_email', 'Teacher Email', 'teacher', 'Teacher'],
    academic_year: ['academic_year', 'Academic Year', 'year', 'Year']
  },
  students: {
    student_id: ['student_id', 'Student ID', 'id', 'ID', 'student_number'],
    full_name: ['full_name', 'Full Name', 'name', 'Name', 'student_name'],
    grade: ['grade', 'Grade', 'GRADE'],
    level: ['level', 'Level', 'LEVEL', 'class_level', 'performance_level'],
    track: ['track', 'Track', 'TRACK', 'stream'],
    class_name: ['class_name', 'Class Name', 'class', 'Class']
  },
  scores: {
    student_id: ['student_id', 'Student ID', 'id', 'ID'],
    exam_name: ['exam_name', 'Exam Name', 'exam', 'Exam'],
    assessment_code: ['assessment_code', 'Assessment Code', 'assessment', 'code'],
    score: ['score', 'Score', 'SCORE', 'marks', 'grade'],
    entered_by_email: ['entered_by_email', 'Teacher Email', 'teacher', 'entered_by']
  }
} as const