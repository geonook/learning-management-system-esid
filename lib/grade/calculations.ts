/**
 * Grade Calculation Functions - Single Source of Truth
 * 
 * Rules:
 * - Only count scores > 0
 * - If all scores are 0 or null, return null
 * - Semester grade rounded to 2 decimal places
 * - Formula: (Formative × 0.15 + Summative × 0.2 + Final × 0.1) ÷ 0.45
 */

import { 
  ScoresMap, 
  ScoreValue, 
  GradeCalculationInput, 
  GradeCalculationResult,
  FORMATIVE_CODES,
  SUMMATIVE_CODES,
  FINAL_CODES 
} from './types'

/**
 * Calculate formative assessment average (FA1-FA8)
 * Only includes scores > 0
 */
export function calcFormativeAvg(scores: ScoresMap): number | null {
  const formativeScores = FORMATIVE_CODES
    .map(code => scores[code])
    .filter((score): score is number => score !== null && score > 0)
  
  if (formativeScores.length === 0) {
    return null
  }
  
  const sum = formativeScores.reduce((acc, score) => acc + score, 0)
  return Number((sum / formativeScores.length).toFixed(2))
}

/**
 * Calculate summative assessment average (SA1-SA4)
 * Only includes scores > 0
 */
export function calcSummativeAvg(scores: ScoresMap): number | null {
  const summativeScores = SUMMATIVE_CODES
    .map(code => scores[code])
    .filter((score): score is number => score !== null && score > 0)
  
  if (summativeScores.length === 0) {
    return null
  }
  
  const sum = summativeScores.reduce((acc, score) => acc + score, 0)
  return Number((sum / summativeScores.length).toFixed(2))
}

/**
 * Get final exam score
 * Returns null if score is null or 0
 */
export function getFinalScore(scores: ScoresMap): number | null {
  const finalScore = scores.FINAL
  return (finalScore !== null && finalScore > 0) ? finalScore : null
}

/**
 * Calculate semester grade using weighted formula
 * Formula: (Formative × 0.15 + Summative × 0.2 + Final × 0.1) ÷ 0.45
 * Returns null if insufficient data
 */
export function calcSemesterGrade(scores: ScoresMap): number | null {
  const formativeAvg = calcFormativeAvg(scores)
  const summativeAvg = calcSummativeAvg(scores)
  const finalScore = getFinalScore(scores)
  
  // Need at least one component to calculate semester grade
  if (!formativeAvg && !summativeAvg && !finalScore) {
    return null
  }
  
  let weightedSum = 0
  let totalWeight = 0
  
  // Add formative component if available
  if (formativeAvg !== null) {
    weightedSum += formativeAvg * 0.15
    totalWeight += 0.15
  }
  
  // Add summative component if available
  if (summativeAvg !== null) {
    weightedSum += summativeAvg * 0.2
    totalWeight += 0.2
  }
  
  // Add final component if available
  if (finalScore !== null) {
    weightedSum += finalScore * 0.1
    totalWeight += 0.1
  }
  
  // Calculate final grade (normalize by actual total weight, not 0.45)
  const semesterGrade = weightedSum / totalWeight
  return Number(semesterGrade.toFixed(2))
}

/**
 * Complete grade calculation with metadata
 * Returns all calculated values plus usage statistics
 */
export function calculateGrades(input: GradeCalculationInput): GradeCalculationResult {
  const { scores } = input
  
  const formativeAvg = calcFormativeAvg(scores)
  const summativeAvg = calcSummativeAvg(scores)
  const semesterGrade = calcSemesterGrade(scores)
  
  // Calculate usage statistics
  const formativeScoresUsed = FORMATIVE_CODES
    .filter(code => scores[code] !== null && scores[code]! > 0).length
  
  const summativeScoresUsed = SUMMATIVE_CODES
    .filter(code => scores[code] !== null && scores[code]! > 0).length
  
  const finalScoreUsed = scores.FINAL !== null && scores.FINAL > 0
  
  const totalScoresUsed = formativeScoresUsed + summativeScoresUsed + (finalScoreUsed ? 1 : 0)
  
  return {
    formativeAvg,
    summativeAvg,
    semesterGrade,
    totalScoresUsed,
    formativeScoresUsed,
    summativeScoresUsed,
    finalScoreUsed
  }
}

/**
 * Utility function to validate if a score is valid (not null and > 0)
 */
export function isValidScore(score: ScoreValue): score is number {
  return score !== null && score > 0
}

/**
 * Get all valid scores from a ScoresMap
 */
export function getValidScores(scores: ScoresMap): { code: string, score: number }[] {
  return Object.entries(scores)
    .filter(([, score]) => isValidScore(score))
    .map(([code, score]) => ({ code, score: score! }))
}