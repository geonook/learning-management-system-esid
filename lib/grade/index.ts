/**
 * Grade Calculation Library - Single Source of Truth
 * 
 * This module exports all grade calculation functions and types.
 * Use these functions for ALL grade calculations across the application.
 * 
 * Key Rules:
 * - Only numerical scores (no letter grades or 等第)
 * - Only count scores > 0 in calculations
 * - Consistent rounding to 2 decimal places
 * - Same logic used in frontend and backend
 */

// Export all calculation functions
export {
  calcFormativeAvg,
  calcSummativeAvg,
  getFinalScore,
  calcSemesterGrade,
  calculateGrades,
  isValidScore,
  getValidScores
} from './calculations'

// Export all types and schemas
export {
  ASSESSMENT_CODES,
  FORMATIVE_CODES,
  SUMMATIVE_CODES,
  FINAL_CODES,
  AssessmentCodeSchema,
  ScoreSchema,
  ScoresMapSchema,
  GradeCalculationInputSchema
} from './types'

export type {
  AssessmentCode,
  ScoreValue,
  ScoresMap,
  GradeCalculationInput,
  GradeCalculationResult,
  AssessmentDisplayOverride
} from './types'

// Export display name utilities
export { resolveDisplayName } from './display-names'