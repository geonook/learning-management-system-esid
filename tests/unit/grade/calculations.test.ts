import { describe, it, expect, test } from 'vitest'
import {
  calcFormativeAvg,
  calcSummativeAvg,
  getFinalScore,
  calcSemesterGrade,
  calculateGrades,
  isValidScore,
  getValidScores
} from '@/lib/grade/calculations'
import { ScoresMap } from '@/lib/grade/types'

describe('Grade Calculations', () => {
  describe('calcFormativeAvg', () => {
    test('calculates average of valid FA scores > 0', () => {
      const scores: ScoresMap = {
        FA1: 80, FA2: 90, FA3: 70, FA4: null, FA5: 0, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: null
      }
      expect(calcFormativeAvg(scores)).toBe(80.0) // (80+90+70)/3 = 80
    })

    test('returns null when all FA scores are null or 0', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: 0, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: null
      }
      expect(calcFormativeAvg(scores)).toBeNull()
    })

    test('excludes 0 scores from calculation', () => {
      const scores: ScoresMap = {
        FA1: 90, FA2: 0, FA3: 80, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: null
      }
      expect(calcFormativeAvg(scores)).toBe(85.0) // (90+80)/2 = 85
    })
  })

  describe('calcSummativeAvg', () => {
    test('calculates average of valid SA scores > 0', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: null, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: 85, SA2: 95, SA3: null, SA4: 0,
        FINAL: null
      }
      expect(calcSummativeAvg(scores)).toBe(90.0) // (85+95)/2 = 90
    })

    test('returns null when all SA scores are null or 0', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: null, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: 0, SA3: null, SA4: null,
        FINAL: null
      }
      expect(calcSummativeAvg(scores)).toBeNull()
    })
  })

  describe('getFinalScore', () => {
    test('returns final score when > 0', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: null, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: 88.5
      }
      expect(getFinalScore(scores)).toBe(88.5)
    })

    test('returns null when final score is 0', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: null, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: 0
      }
      expect(getFinalScore(scores)).toBeNull()
    })

    test('returns null when final score is null', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: null, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: null
      }
      expect(getFinalScore(scores)).toBeNull()
    })
  })

  describe('calcSemesterGrade', () => {
    test('calculates weighted semester grade with all components', () => {
      const scores: ScoresMap = {
        FA1: 80, FA2: 90, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null, // avg = 85
        SA1: 88, SA2: 92, SA3: null, SA4: null, // avg = 90
        FINAL: 94
      }
      // Weighted: (85*0.15 + 90*0.2 + 94*0.1) / 0.45 = (12.75 + 18 + 9.4) / 0.45 = 89.22
      expect(calcSemesterGrade(scores)).toBe(89.22)
    })

    test('calculates with partial components', () => {
      const scores: ScoresMap = {
        FA1: 80, FA2: 90, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null, // avg = 85
        SA1: null, SA2: null, SA3: null, SA4: null, // no SA scores
        FINAL: 94
      }
      // Only FA and Final: (85*0.15 + 94*0.1) / 0.25 = (12.75 + 9.4) / 0.25 = 88.6
      expect(calcSemesterGrade(scores)).toBe(88.6)
    })

    test('returns null when no valid scores', () => {
      const scores: ScoresMap = {
        FA1: null, FA2: 0, FA3: null, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: null
      }
      expect(calcSemesterGrade(scores)).toBeNull()
    })
  })

  describe('calculateGrades - complete integration', () => {
    test('returns complete calculation result with metadata', () => {
      const input = {
        scores: {
          FA1: 80, FA2: 90, FA3: 0, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
          SA1: 88, SA2: 92, SA3: null, SA4: null,
          FINAL: 94
        } as ScoresMap,
        studentId: 'student-1',
        classId: 'class-1'
      }

      const result = calculateGrades(input)

      expect(result).toEqual({
        formativeAvg: 85.0, // (80+90)/2
        summativeAvg: 90.0, // (88+92)/2
        semesterGrade: 89.22, // weighted calculation
        totalScoresUsed: 5, // FA1, FA2, SA1, SA2, FINAL
        formativeScoresUsed: 2, // FA1, FA2
        summativeScoresUsed: 2, // SA1, SA2
        finalScoreUsed: true
      })
    })

    test('handles mixed score scenarios correctly', () => {
      const input = {
        scores: {
          FA1: 0, FA2: null, FA3: 95, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
          SA1: null, SA2: 0, SA3: null, SA4: null,
          FINAL: 85
        } as ScoresMap,
        studentId: 'student-2',
        classId: 'class-2'
      }

      const result = calculateGrades(input)

      expect(result).toEqual({
        formativeAvg: 95.0, // Only FA3 counts
        summativeAvg: null, // No valid SA scores
        semesterGrade: 87.6, // (95*0.15 + 85*0.1) / 0.25
        totalScoresUsed: 2, // FA3, FINAL
        formativeScoresUsed: 1,
        summativeScoresUsed: 0,
        finalScoreUsed: true
      })
    })
  })

  describe('utility functions', () => {
    test('isValidScore correctly identifies valid scores', () => {
      expect(isValidScore(85.5)).toBe(true)
      expect(isValidScore(0)).toBe(false)
      expect(isValidScore(null)).toBe(false)
      expect(isValidScore(100)).toBe(true)
      expect(isValidScore(0.1)).toBe(true)
    })

    test('getValidScores returns only valid score entries', () => {
      const scores: ScoresMap = {
        FA1: 85, FA2: 0, FA3: null, FA4: 90, FA5: null, FA6: null, FA7: null, FA8: null,
        SA1: null, SA2: null, SA3: null, SA4: null,
        FINAL: 88.5
      }

      const validScores = getValidScores(scores)
      expect(validScores).toEqual([
        { code: 'FA1', score: 85 },
        { code: 'FA4', score: 90 },
        { code: 'FINAL', score: 88.5 }
      ])
    })
  })

  // Snapshot tests for regression prevention
  describe('snapshot tests', () => {
    test('complete score calculation snapshot', () => {
      const input = {
        scores: {
          FA1: 85.5, FA2: 78.0, FA3: 92.5, FA4: null, FA5: null, FA6: null, FA7: null, FA8: null,
          SA1: 88.0, SA2: 91.5, SA3: null, SA4: null,
          FINAL: 89.0
        } as ScoresMap,
        studentId: 'snapshot-test-student',
        classId: 'snapshot-test-class'
      }

      const result = calculateGrades(input)
      expect(result).toMatchSnapshot()
    })

    test('partial score calculation snapshot', () => {
      const input = {
        scores: {
          FA1: null, FA2: 95.0, FA3: null, FA4: 87.5, FA5: null, FA6: null, FA7: null, FA8: null,
          SA1: 90.0, SA2: null, SA3: null, SA4: null,
          FINAL: null
        } as ScoresMap,
        studentId: 'partial-scores-student',
        classId: 'partial-scores-class'
      }

      const result = calculateGrades(input)
      expect(result).toMatchSnapshot()
    })

    test('zero and null scores handling snapshot', () => {
      const input = {
        scores: {
          FA1: 0, FA2: 95.0, FA3: null, FA4: 87.5, FA5: null, FA6: null, FA7: null, FA8: null,
          SA1: 90.0, SA2: 0, SA3: null, SA4: null,
          FINAL: 85.5
        } as ScoresMap,
        studentId: 'mixed-scores-student',
        classId: 'mixed-scores-class'
      }

      const result = calculateGrades(input)
      expect(result).toMatchSnapshot()
    })
  })
})