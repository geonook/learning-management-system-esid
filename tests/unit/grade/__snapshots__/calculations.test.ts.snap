// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Grade Calculations > snapshot tests > complete score calculation snapshot 1`] = `
{
  "finalScoreUsed": true,
  "formativeAvg": 85.33,
  "formativeScoresUsed": 3,
  "semesterGrade": 88.11,
  "summativeAvg": 89.75,
  "summativeScoresUsed": 2,
  "totalScoresUsed": 6,
}
`;

exports[`Grade Calculations > snapshot tests > partial score calculation snapshot 1`] = `
{
  "finalScoreUsed": false,
  "formativeAvg": 91.25,
  "formativeScoresUsed": 2,
  "semesterGrade": 90.54,
  "summativeAvg": 90,
  "summativeScoresUsed": 1,
  "totalScoresUsed": 3,
}
`;

exports[`Grade Calculations > snapshot tests > zero and null scores handling snapshot 1`] = `
{
  "finalScoreUsed": true,
  "formativeAvg": 91.25,
  "formativeScoresUsed": 2,
  "semesterGrade": 89.42,
  "summativeAvg": 90,
  "summativeScoresUsed": 1,
  "totalScoresUsed": 4,
}
`;
